### ISSUE TITLE
Unify checking for libnccl in setup.py and build_all.sh

### ISSUE THREAD

**Issue by csarofeen** [pytorch/pytorch#2375]
In order to install pytorch with a user installed nccl the following steps are needed:

to satisfy setup.py:
Add .../nccl/build/lib in /etc/ld.so.conf
run ldconfig

to satisfy build_all.sh:
export CMAKE_PREFIX_PATH=".../nccl/build":$CMAKE_PREFIX_PATH

to satisfy the linker:
export LIBRARY_PATH=".../nccl/build/lib/":$LIBRARY_PATH

is there a way we could unify these steps to make the process more straight forward?

Another problem may arise if the nccl is installed without the header. It may be worth having setup.py try to locate the header and give a descriptive error if it is not found with the library.

**Comment by tpankaj** [comment]
Even with these steps, I'm still unable to compile your Cuda9 PR #2263. It appears to be wanting the nccl source tree.
```
CMake Error at /home/ubuntu/anaconda3/share/cmake-3.6/Modules/ExternalProject.cmake:1924 (message):
  No download info given for 'nccl_external' and its source directory:

   /home/ubuntu/pytorch/torch/lib/gloo/third-party/nccl

  is not an existing non-empty directory.  Please specify one of:

   * SOURCE_DIR with an existing non-empty directory
   * URL
   * GIT_REPOSITORY
   * HG_REPOSITORY
   * CVS_REPOSITORY and CVS_MODULE
   * SVN_REVISION
   * DOWNLOAD_COMMAND
```

**Comment by csarofeen** [comment]
What's the output of:
```echo $CMAKE_PREFIX_PATH```
and
```ldconfig -p | grep libnccl```
?

**Comment by tpankaj** [comment]
```
ubuntu@aws-box:~$ echo $CMAKE_PREFIX_PATH
.../nccl/build:
ubuntu@aws-box:~$ ldconfig -p | grep libnccl
        libnccl.so.2 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libnccl.so.2
        libnccl.so (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libnccl.so
```

**Comment by csarofeen** [comment]
Sorry, when I put ... I meant ```path_to_nccl/build/```.
```CMAKE_PREFIX_PATH``` needs to point to a directory where there is an ```include``` dir with nccl.h and a ```lib``` with  the .so files.
If you compiled nccl from source there may be ```build``` dir where you installed the code from. If so that's the dir ```CMAKE_PREFIX_PATH``` should be pointed to.

**Comment by tpankaj** [comment]
I see, that makes much more sense.

**Comment by tpankaj** [comment]
It builds properly now.

**Comment by TeslasGhost** [comment]
@tpankaj - did you install nccl v1 or nccl v2? I cannot seem to resolve the dependencies with nccl2

**Comment by soumith** [comment]
i'm looking into / working on this.

**Comment by soumith** [comment]
fixed via #2853
